src/qiskit_qasm3_import/state.py

```diff
@@ -0,0 +1,205 @@
+        """Set the addressing mode to physical. On success return `True`, otherwise raise an
+        exception."""
+        if self._state == self.VIRTUAL:
+            raise_from_node(
+                node,
+                "Physical qubit referenced in virtual addressing mode. Mixing "
+                "modes not currently supported.",
+            )
+        self._state = self.PHYSICAL
+
+    def set_virtual_mode(self, node):
+        """Set the addressing mode to virtual.
+        On success return `True`, otherwise raise an exception."""
+        if self._state == self.PHYSICAL:
+            raise_from_node(
+                node,
+                "Virtual qubit declared in physical addressing mode. Mixing modes not currently "
+                "supported.",
+            )
+        self._state = self.VIRTUAL
+
+
+class SymbolTable:
+    __slots__ = ("symbols", "scope", "base", "global_symbols")
+
+    def __init__(self, scope: Scope, base: Optional["SymbolTable"] = None):
+        self.symbols = {}
+        self.scope = scope
+        self.base = base
+        self.global_symbols = base.global_symbols if base is not None else _BUILTINS.copy()
+
+    def __contains__(self, name: str):
+        return name in (self.symbols, self.global_symbols) or (
+            self.base is not None and name in self.base.symbols
+        )
+
+    def get(self, name: str, node=None):
+        """Lookup symbol `name`
+
+        Return `None` if no symbol is found. Raise an exception if the symbol is
+        inaccessible in the scope of this table.
+        """
+        if (symbol := self.symbols.get(name, None)) is not None:
+            return symbol
+        if self.base is not None:
+            if (symbol := self.base.get(name, None)) is None:
+                return None
+            return self._check_visible(symbol, node)
+        return self.global_symbols.get(name, None)
+
+    def _check_visible(self, symbol, node):
+        if (
+            symbol.scope is Scope.BUILTIN
+            or self.scope is not Scope.GATE
+            or symbol.scope is not Scope.GLOBAL
+        ):
+            return symbol
+        if self.scope is Scope.GATE:
+            if (
+                isinstance(symbol.type, types.Gate)
+                or isinstance(
+                    symbol.type, (types.Int, types.Uint, types.Float, types.Angle, types.Duration)
+                )
+                and symbol.type.const
+            ):
+                return symbol
+        if isinstance(symbol.type, types.HardwareQubit):
+            raise PhysicalQubitInGateError(node.name, node)
+        raise_from_node(node, f"Symbol {symbol.name} is not visible in the scope of a gate")
+
+    def insert(self, symbol):
+        target = self.global_symbols if symbol.type == types.HardwareQubit() else self.symbols
+        target[symbol.name] = symbol
+
+    def globals(self):
+        if self.global_symbols is None:
+            return []
+        return self.global_symbols.values()
```

Comment:
In the current situation, I think `self.global_symbols` is always not `None`, though see above comment since I'm not sure it should be a raw dict.

Reply:
Yes that's correct. This must be a relic.

Reply:
If `global_symbols` is not a raw `dict` what should it be?

Reply:
Fixed in 1cd0d3a . But we may redesign this anyway.